zerver/actions/presence.py

```diff
@@ -77,39 +81,62 @@ def do_update_user_presence(
 ) -> None:
     client = consolidate_client(client)
 
+    # TODO: While we probably DO want creating an account to
+    # automatically create a first `UserPresence` object with
+    # last_connected_time and last_active_time as the current time,
+    # our presence tests don't understand this, and it'd be perhaps
+    # wrong for some cases of account creation via the API.  So we may
+    # want a "never" value here as the default.
     defaults = dict(
-        timestamp=log_time,
-        status=status,
+        # Given that these are defaults for creation of a UserPresence row
+        # if one doesn't yet exist, the most sensible way to do this
+        # is to set both last_active_time and last_connected_time
+        # to log_time.
+        last_active_time=log_time,
+        last_connected_time=log_time,
         realm_id=user_profile.realm_id,
     )
+    if status == UserPresence.LEGACY_STATUS_IDLE_INT:
+        # If the presence entry for the user is just to be created, and
+        # we want it to be created as idle, then there needs to be an appropriate
+        # offset between last_active_time and last_connected_time, since that's
+        # what the event-sending system calculates the status based on, via
+        # format_legacy_presence_dict.
+        defaults["last_active_time"] = log_time - datetime.timedelta(
+            seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS + 1
+        )
 
     (presence, created) = UserPresence.objects.get_or_create(
         user_profile=user_profile,
-        client=client,
         defaults=defaults,
     )
 
-    stale_status = (log_time - presence.timestamp) > datetime.timedelta(minutes=1, seconds=10)
-    was_idle = presence.status == UserPresence.IDLE
-    became_online = (status == UserPresence.ACTIVE) and (stale_status or was_idle)
+    time_since_last_active = log_time - presence.last_active_time
+    now_online = time_since_last_active > datetime.timedelta(
+        seconds=settings.OFFLINE_THRESHOLD_SECS - 10
+    )
```

Comment:
Hmm, is this `-10` logic correct? I think the idea is we want to immediately notify clients if they would have previously shown the user as offline. It's weird -- if we did exactly OFFLINE_THRESHOLD_SECONDS, that'd be wrong in that we might fail to notify clients that were about to start displaying the user as offline. At the same time, we only want this code path to send a presence update via the events system if the user's state has actually changed.

Do we want `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL`, in other words, "if a client might not otherwise display you as offline before the next time they ping the server"? I worry that might be so small that this code path triggers incorrectly really often.

Reply:
> Do we want `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL`, in other words, "if a client might not otherwise display you as offline before the next time they ping the server"? I worry that might be so small that this code path triggers incorrectly really often.

`OFFLINE_THRESHOLD_SECONDS` is 5 minutes though, and `ACTIVE_PING_INTERVAL_MS = 50 * 1000` (50 seconds) so that delta is actually 4 minutes 10 seconds so not short at all. So perhaps that would be fine, unless I'm thinking about this wrong and that is too large instead?

Reply:
OK if `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL - 10` is 4 minutes, that's probably very reasonable for this -- it should mean that a client only triggers this if it's missed like 4 pings in a row, which is a good definitely for "was previously offline".

Oh, but the frontend has `OFFLINE_THRESHOLD_SECS=140`; and we had that hardcoded in `message_send.py` too. I think this is all good IF we sort out that discrepancy. I like the idea of this being a setting long-term, since it's a thing we can maybe tune in Zulip Cloud during an incident or something, so we may want to put it into the `/server_settings` API at some point. 

```
$ git grep OFFLINE_THRESHOLD_SECS
docs/subsystems/presence.md:  useful sample code; the `OFFLINE_THRESHOLD_SECS` check is critical
frontend_tests/node_tests/presence.js:const OFFLINE_THRESHOLD_SECS = 140;
frontend_tests/node_tests/presence.js:        active_timestamp: now - OFFLINE_THRESHOLD_SECS / 2,
frontend_tests/node_tests/presence.js:        active_timestamp: now - OFFLINE_THRESHOLD_SECS * 2,
frontend_tests/node_tests/presence.js:        idle_timestamp: now - OFFLINE_THRESHOLD_SECS / 2,
frontend_tests/node_tests/presence.js:    const recent = now + 1 - OFFLINE_THRESHOLD_SECS;
frontend_tests/node_tests/presence.js:    const a_while_ago = now - OFFLINE_THRESHOLD_SECS * 2;
static/js/presence.js:const OFFLINE_THRESHOLD_SECS = 140;
static/js/presence.js:    if (age(active_timestamp) < OFFLINE_THRESHOLD_SECS) {
static/js/presence.js:    if (age(idle_timestamp) < OFFLINE_THRESHOLD_SECS) {
zerver/actions/message_send.py:    OFFLINE_THRESHOLD_SECS = 140
zerver/actions/message_send.py:    recent = timezone_now() - datetime.timedelta(seconds=OFFLINE_THRESHOLD_SECS)
zerver/actions/user_settings.py:                seconds=settings.OFFLINE_THRESHOLD_SECS + 120
zerver/tests/test_presence.py:        with self.settings(OFFLINE_THRESHOLD_SECS=1):
zerver/views/presence.py:    if aggr_status_duration > settings.OFFLINE_THRESHOLD_SECS:
zproject/default_settings.py:OFFLINE_THRESHOLD_SECS = 5 * 60
```

It looks like the randomly choosing between those two in values in different places; but I think it may be the 140s is the operative one today, in that `status_from_raw` mostly looks at the timestamps, not the server `offline` flag...

Reply:
OK, I took the time to analyze the history here, and posted a plan in the thread here:

The proposal in https://chat.zulip.org/#narrow/stream/3-backend/topic/UserPresence.20system.20rewrite/near/1506240

With that plan, this line can be `settings.OFFLINE_THRESHOLD_SECS - ACTIVE_PING_INTERVAL - 10` which would be 130s -- long enough that we'd need to have missed two pings in a row; that seems safe and correct to me. We may want to do something to monitor that this doesn't happen excessively often; but that should be visible in the event logs.